accepted/2025/cli-acquisition-tool.md

```diff
@@ -0,0 +1,224 @@
+- **Self-maintenance**: Keep the bootstrapper tool itself up to date
+
+### Non-Goals
+
+- **Package Manager Integration/Global Installations**: Will not support installers beyond .zip/.tar installs initially
+- **Replace dotnet.exe**: The bootstrapper will not be dotnet.exe itself
+- **NuGet/Tools Support**: Tools are not in scope for `dnup` - instead they would belong to a NAOT `dnx` or NAOT'd `dotnet` CLI
+- **SDK Size Reduction**: Size reduction will be tackled separately, focused on deduplication and factoring out slices of functionality
+
+## Stakeholders and Reviewers
+
+- **.NET SDK Team**: Core implementation and CLI design
+- **VS Code C# Extension Team**: Integration scenarios and user experience
+- **Visual Studio Team**: Ensuring compatibility with existing VS workflows
+- **Security Team**: Signature verification and SDL compliance
+- **.NET Foundation**: Community feedback and adoption strategy
+- **Documentation Team**: Getting started guides and migration documentation
+
+## Design
+
+### Bootstrapper CLI Design
+
+The tool will follow the CLI design outlined in the [related PR](https://github.com/dotnet/designs/pull/335) and will include commands like
+
+* install specific versions of the SDK tooling (10.0.100)
+* install logically-named versions of the SDK tooling (10, LTS, etc)
+* list the installed SDKs
+* uninstall specific versions of the SDK tooling
+* install the versions of components required by a given repo (as specified by global.json)
+* provide update checks/out-of-date information to end users about the installs they have
+
+### Initial setup
+
+On first-run, `dnup-init` will configure the users' environment to support user-local hives.
+* `DOTNET_ROOT` will be set to point to the `dnup` user-local root,
+* `PATH` will have the `dnup` user-local root appended to it,
+* on Windows `dnup-init` will request to elevate to set a Windows Registry key that will prevent global .NET installers from setting the global .NET Install location on the System `PATH` (which would always overwrite the user `PATH`).
+  * this behavior _may_ be optional - we may choose to only run it by default if the user has a global .NET install already
+
+This configuration will be necessity be shell-specific, so the tool will need to detect the user's shell and modify the appropriate profile file (e.g. `.bashrc`, `.zshrc`, etc.) to set the `DOTNET_ROOT` and `PATH` variables.
+We have some prior art in the `dotnet` CLI codebase for detecting the shell that is in use, which can be reused here.
+The different shells (and to some extent different OS platforms) have different ways of distributing such changes, so the tool will need to know about those to be a bit opinionated about how it makes these changes.
+
+This shell-specific configuration will need to be independently run via a command, because the user may choose a different shell after the initial setup.
+
+### Install locations
```

Comment:
I don't know if this is an explicit goal or not, but something myself and other folks have struggled with is juggling multiple architectures for "dotnet". Windows has a few different architecture flavors for x86, x64, and ARM64. Will, or how, will this tool make it easier to juggle the "I need .NET, no, the other one" problem?

Reply:
We haven't put _too_ much thought into this, but the current best-guess would be since we're already in the business of managing 'hives', we'd track separate hives per-arch. Install/update commands would let you specify which hive (default would be the current arch, flags to update across all, etc). There would be a specific command to 'switch' the active hive, and a way to tell which hive you were currently in.  Similar in spirit to how `nvm use` lets you set up a shell to work on a specific version of node.  Thoughts on this general approach? Happy to push it into the doc.

Reply:
> f I need to do development against an IA32-only component on AMD64, what would the developer inner loop experience be like if there was no 32-bit SDK, only runtime?

Add `<RuntimeIdentifier>win-x86</RuntimeIdentifier>` to your .csproj file and your app will be compiled and run as 32-bit with .NET 64-bit SDK. It is very similar to how e.g. mobile development works. It is also how a people often target linux-arm32. Native .NET SDK is incredibly slow on arm32 devices.

If you want the project to be arch neutral and you just want to have a one-off terminal window where it builds and runs as 32-bit, you can set RuntimeIdentifier=win-x86 in that window.

For one-off run as 32-bit, you can use `-a` argument, e.g. `dotnet run -a x86`.